### PR TITLE
UPNP Port Forwarding Fail On Local IP Change#698

### PR DESCRIPTION
**Pull request by apoorv-2204**
# Description

Attempt to resolve  upnp_port_forwarding_fail_on_local_ip_change#698.
- If there is no redirection -> continue and forward port -> `Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP.open(port)`
- If there is a redirection on the same local ip -> continue and forward port (to update time limit)
   miniupnpc duration set to 0 as for mapping for a lifetime
- If there is a redirection on a different local ip -> remove the existing redirection then continue and forward port
   Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP, we expect this to fail and capture conflict mapping entry
error .revoke that port and assigns to the node

Fixes 

- [ ] #698 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?
- When the port forwarding fails , we capture the error, and remove port rule , and redo the rule.
 Screenshot shows working of code, trying to imitate that condition ,we expect same when node fails to port forward.

proof

# Checklist:

- My code not follows the style guidelines of this project
- I have not performed a self-review of my own code
- I have commented my code, particularly in easy-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate 20 new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 1]
```diff
@@ -43,8 +43,31 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
     end
   end
 
+  @spec do_open_port(:inet.ip_address(), non_neg_integer()) :: :ok | :error
   defp do_open_port(local_ip, port) do
-    opts = [
+    with {reason, status} when status != 0 <- System.cmd(@upnpc, map_query(local_ip, port)),
```
Comment (bchamagne) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:48]: I find this pattern very hard to read

Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:48]: I think using the happy path firstly, we would make more sense.
The error branching should be done on the else, I think.

**Review by bchamagne — APPROVED** [review]
Just read the code, I did not test it.

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex (hunk 1 above)
Comment (samuelmanzanera) [reply lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:48]: We could use something like this:
```elixir
case System.cmd(@upnpc, map_query(local_ip, port)) do
  {_, 0} ->
    :ok 
  {reason, status} ->
    reason
    |> parse_reason()
    |> handle_error()
 end
```

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 2]
```diff
@@ -54,18 +77,44 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
...
-    case System.cmd(@upnpc, opts) do
-      {_, 0} ->
-        :ok
+  @spec parse_reason(String.t()) ::
+          {:error, :conflict_in_mapping_entry | :unknown}
+  defp parse_reason(reason) do
+    # upon more condtions , refactor with cond do end
+    if Regex.scan(~r/ConflictInMappingEntry/, reason, capture: :all) != [] do
+      Logger.warning("Port is employed to another host.")
+      {:error, :conflict_in_mapping_entry}
```
Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:92]: The error tuple is more used to help to branch between success and failure.
Here you just want to format error's reason.
I think you could return simplify the error's reason, so you could easily use this function in a pipeline.

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 3]
```diff
@@ -54,18 +77,44 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
...
+  end
 
-      {reason, _status} ->
-        Logger.debug(reason)
-        :error
+  @spec handle_error(:conflict_in_mapping_entry | :unknown, :inet.ip_address(), non_neg_integer()) ::
+          :ok | :error
+  defp handle_error(:conflict_in_mapping_entry, _, port) do
+    case System.cmd(@upnpc, revoke_query(port)) do
+      {_, 0} -> :ok
```
Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:102]: I think you could leverage the reopening in the error's handling.

**Review by samuelmanzanera — CHANGES_REQUESTED** [review]

**Comment by apoorv-2204** [comment]
image

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 4]
```diff
@@ -43,8 +43,30 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
     end
   end
 
-  defp do_open_port(local_ip, port) do
-    opts = [
+  @spec do_open_port(:inet.ip_address(), non_neg_integer()) :: :ok | :error
+  def do_open_port(local_ip, port, retires \\ 2)
```
Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:47]: You might keep it private

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 5]
```diff
@@ -43,8 +43,30 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
...
+        do_open_port(local_ip, port, retires - 1)
+    end
+  end
+
+  @protocol "tcp"
+  @spec map_query(:inet.ip_address(), non_neg_integer()) :: [String.t()]
+  defp map_query(local_ip, port) do
+    [
+      # description
+      "-e",
```
Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:68]: Last time we got some issue with this parameter. Need to check with @Neylix

**Review by samuelmanzanera — CHANGES_REQUESTED** [review]

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 6]
```diff
@@ -43,8 +43,27 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
...
   end
 
-  defp do_open_port(local_ip, port) do
-    opts = [
+  @spec do_open_port(:inet.ip_address(), non_neg_integer()) :: :ok | :error
+  defp do_open_port(local_ip, port, retires \\ 2)
+
+  defp do_open_port(_local_ip, _port, 0), do: :error
+
+  defp do_open_port(local_ip, port, retires) do
```
Comment (bchamagne) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:51]: typo: `retries`

ps: very good idea

File: lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex [hunk 7]
```diff
@@ -43,8 +43,27 @@ defmodule Archethic.Networking.IPLookup.NATDiscovery.MiniUPNP do
...
+    case System.cmd(@upnpc, map_query(local_ip, port)) do
+      {_, 0} ->
+        :ok
+
+      {reason, status} when status != 0 ->
```
Comment (samuelmanzanera) [inline lib/archethic/networking/ip_lookup/nat_discovery/miniupnp.ex:56]: I think you can omit the status check

**Review by samuelmanzanera — COMMENTED** [review]
Code seems ok

**Review by Neylix — APPROVED** [review]

**Pull request merged**